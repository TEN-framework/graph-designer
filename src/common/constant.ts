const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL || "http://localhost"
const PORT = process.env.NEXT_PUBLIC_PORT || 49483
const PREFIX = "/api/dev-server/v1"


export const DEFAULT_APP = "localhost"
export const DEFAULT_EXTENTION_GROUP = "default"
export const API_URL = `${BASE_URL}:${PORT}${PREFIX}`
