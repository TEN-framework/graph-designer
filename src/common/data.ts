export class EditorData {
  // extensionGroup =>  {
  //  [extensionName]: nodeId
  // }
  nodeMap: Map<string, { [extensionName: string]: string }> = new Map()
  nodeId: number = 1
  edgeId: number = 1

  constructor() { }

  genEdgeId() {
    return `${this.edgeId++}`
  }

  genNodeId() {
    return `${this.nodeId++}`
  }

  getNodeId(extensionGroup: string, extensionname: string) {
    let data = this.nodeMap.get(extensionGroup)
    if (data) {
      if (!data[extensionname]) {
        throw new Error(
          `Invalid extension Node: ${extensionname}, not found in extensionGroup: ${extensionGroup}`,
        )
      }
      return data[extensionname]
    }
    throw new Error(
      `Invalid extensionGroup: ${extensionGroup}, not found in nodeMap`,
    )
  }

  delNode(extensionGroup: string, extensionname: string) {
    let data = this.nodeMap.get(extensionGroup)
    if (data) {
      delete data[extensionname]
    }
  }

  saveNodeId(extensionGroup: string, extensionname: string, nodeId: string) {
    let data = this.nodeMap.get(extensionGroup)
    if (!data) {
      data = {}
      this.nodeMap.set(extensionGroup, data)
    }
    data[extensionname] = nodeId
  }

  clear() {
    this.nodeMap.clear()
    this.nodeId = 1
    this.edgeId = 1
  }
}


export const editorData = new EditorData()

if (process.env.NODE_ENV === "development") {
  if (typeof window !== "undefined") {
    // @ts-ignore
    window.editorData = editorData
  }
}
