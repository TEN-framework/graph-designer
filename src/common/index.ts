export * from "./logger"
export * from "./utils"
export * from "./request"
export * from "./hooks"
export * from "./mock"
export * from "./event"
export * from "./constant"
