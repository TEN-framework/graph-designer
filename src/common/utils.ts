import { Edge, Node, XYPosition, Position } from "@xyflow/react"
import {
  IExtension,
  IConnection,
  IConnectionData,
  MsgType,
  InOutData,
  IExtensionNode,
} from "@/types"

let EDGE_ID = 0

const getEdgeId = () => {
  return `${EDGE_ID++}`
}

function _pad(num: number) {
  return num.toString().padStart(2, "0")
}

export const formatTime = (date?: Date) => {
  if (!date) {
    date = new Date()
  }
  const hours = date.getHours()
  const minutes = date.getMinutes()
  const seconds = date.getSeconds()
  const milliseconds = date.getMilliseconds()

  return `${_pad(hours)}:${_pad(minutes)}:${_pad(seconds)}:${_pad(milliseconds)}`
}

export const round = (min: number, max: number) => {
  return Math.floor(Math.random() * (max - min + 1)) + min
}

// DataTest => data_test
export const camelToSnake = (str: string): string => {
  return str.replace(/[A-Z]/g, (match) => "_" + match.toLowerCase()).slice(1)
}

export const sleep = async (ms: number) => {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

// ----------------------- graph ---------------------
export const extensionsToNodes = (
  extensions: IExtension[],
): IExtensionNode[] => {
  return extensions.map((extension, index) => {
    const position = {
      x: index * 250,
      y: (index % 2 == 0 ? 1 : -1) * 100 + round(-130, 130),
    }
    return extensionToNode(extension, { position })
  })
}

export const extensionToNode = (
  extension: IExtension,
  property: {
    position: XYPosition
  },
): IExtensionNode => {
  const { position } = property
  const inputs: InOutData[] = []
  const outputs: InOutData[] = []
  const { api } = extension
  if (api?.cmd_in) {
    api.cmd_in.forEach((input) => {
      inputs.push({ id: input.name, type: "cmd", status: "default" })
    })
  }
  if (api?.cmd_out) {
    api.cmd_out.forEach((output) => {
      outputs.push({ id: output.name, type: "cmd", status: "default" })
    })
  }
  if (api?.data_in) {
    api.data_in.forEach((input) => {
      inputs.push({ id: input.name, type: "data", status: "default" })
    })
  }
  if (api?.data_out) {
    api.data_out.forEach((output) => {
      outputs.push({ id: output.name, type: "data", status: "default" })
    })
  }
  if (api?.pcm_frame_in) {
    api.pcm_frame_in.forEach((input) => {
      inputs.push({ id: input.name, type: "pcm_frame", status: "default" })
    })
  }
  if (api?.pcm_frame_out) {
    api.pcm_frame_out.forEach((output) => {
      outputs.push({ id: output.name, type: "pcm_frame", status: "default" })
    })
  }
  if (api?.img_frame_in) {
    api.img_frame_in.forEach((input) => {
      inputs.push({ id: input.name, type: "img_frame", status: "default" })
    })
  }
  if (api?.img_frame_out) {
    api.img_frame_out.forEach((output) => {
      outputs.push({ id: output.name, type: "img_frame", status: "default" })
    })
  }

  return {
    // TIP: extension.name just a property
    // id should use extension.addon (like: class name)
    id: extension.addon,
    sourcePosition: Position.Right,
    targetPosition: Position.Left,
    position: position,
    type: "extension",
    data: {
      status: "default",
      name: extension.name,
      inputs: inputs,
      outputs: outputs,
      extensionGroup: extension.extension_group,
    },
  }
}

const _connectionDataToEdge = (
  extensionName: string,
  data: IConnectionData[],
): Edge[] => {
  let edges: Edge[] = []
  data.forEach((i) => {
    const { dest = [], name } = i
    dest.forEach((d) => {
      let edg = {
        id: getEdgeId(),
        source: extensionName,
        sourceHandle: `${extensionName}/${name}`,
        target: d.extension,
        targetHandle: `${d.extension}/${name}`,
      } as Edge
      edges.push(edg)
    })
  })

  return edges
}

export const connectionsToEdges = (connections: IConnection[]): Edge[] => {
  const edges: Edge[] = []
  connections.forEach((connection) => {
    const { cmd = [], data = [], pcm_frame = [], extension } = connection
    if (cmd.length) {
      const cmdEdges = _connectionDataToEdge(extension, cmd)
      edges.push(...cmdEdges)
    }
    if (data.length) {
      const dataEdges = _connectionDataToEdge(extension, data)
      edges.push(...dataEdges)
    }
    if (pcm_frame.length) {
      const pcmFrameEdges = _connectionDataToEdge(extension, pcm_frame)
      edges.push(...pcmFrameEdges)
    }
    return edges
  })

  return edges
}

export const handleIdToType = (id: string): MsgType => {
  const data = id.split("/")
  if (!data[1]) {
    throw new Error(`Invalid handle id: ${id}`)
  }
  return data[1] as MsgType
}

export const nodesToExtensions = (
  nodes: IExtensionNode[],
  installedExtensions: IExtension[],
): IExtension[] => {
  return nodes.map((node) => {
    const { id, data } = node
    const { extensionGroup = "default" } = data
    const extension = installedExtensions.find((i) => i.name == id)
    if (!extension) {
      throw new Error(`Invalid extension: ${id}`)
    }
    return {
      ...extension,
      extension_group: extensionGroup,
    }
  })
}

export const edgesToConnections = (edges: Edge[]): IConnection[] => {
  let connections: IConnection[] = []
  const app = "localhost"
  const extension_group = "default"
  for (let edge of edges) {
    const { source, sourceHandle, target, targetHandle } = edge
    const sourceHandleName = sourceHandle?.split("/")[1]
    if (!sourceHandleName) {
      throw new Error(`Invalid source handle: ${sourceHandle}`)
    }
    const curConnection = connections.find((i) => i.extension == source)
    if (curConnection) {
      curConnection.data = curConnection?.data ? curConnection.data : []
      const curData = curConnection.data.find(
        (item) => item.name == sourceHandleName,
      )
      if (curData) {
        curData.dest.push({
          app: "localhost",
          extension_group,
          extension: target,
        })
      } else {
        curConnection.data.push({
          name: sourceHandleName,
          dest: [
            {
              app: "localhost",
              extension_group,
              extension: target,
            },
          ],
        })
      }
    } else {
      connections.push({
        app,
        extension: source,
        extension_group,
        data: [
          {
            name: sourceHandleName,
            dest: [
              {
                app: "localhost",
                extension_group,
                extension: target,
              },
            ],
          },
        ],
      })
    }
  }

  return connections
}
