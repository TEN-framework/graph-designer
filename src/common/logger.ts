import { formatTime } from "./utils"
import { LogLevel } from "../types"

export interface LoggerConfig {
  level?: LogLevel
  prefix?: string
}

class Logger {
  level: LogLevel = LogLevel.ERROR
  prefix?: string = ""
  preTime?: number = 0

  constructor(config: LoggerConfig) {
    const { level, prefix } = config
    if (level !== undefined) {
      this.level = level
    }
    if (prefix) {
      this.prefix = prefix
    }
  }

  setLogLevel(level: LogLevel) {
    this.level = level
  }

  debug(...args: any[]) {
    this.level <= LogLevel.DEBUG &&
      this._log(`${this._genPrefix()}[DEBUG]: `, ...args)
  }

  warn(...args: any[]) {
    this.level <= LogLevel.WARN &&
      this._warn(`${this._genPrefix()}[WARN]: `, ...args)
  }

  error(...args: any[]) {
    this.level <= LogLevel.ERROR &&
      this._err(`${this._genPrefix()}[ERROR]: `, ...args)
  }

  time(...args: any[]) {
    const time = new Date().getTime()
    let cost = 0
    let start = ""
    if (this.preTime) {
      cost = time - this.preTime
    } else {
      start = "start"
    }
    this.preTime = time
    this._log(
      `${this._genPrefix()}[time]:  -------------- cost:${cost}ms ${start} -----------------   \n`,
      ...args,
    )
  }

  timeEnd(...args: any[]) {
    const time = new Date().getTime()
    let cost = 0
    if (this.preTime) {
      cost = time - this.preTime
    }
    this.preTime = 0
    this._log(
      `${this._genPrefix()}[time]:  -------------- cost:${cost}ms end -----------------   \n`,
      ...args,
    )
  }

  //  ---------------------------- private ----------------------------

  private _log(...args: any[]) {
    const [arg, ...other] = args
    console.log(`%c${arg}`, "color:yellow", ...other)
  }

  private _warn(...args: any[]) {
    const [arg, ...other] = args
    console.warn(`%c${arg}`, "color:yellow", ...other)
  }

  private _err(...args: any[]) {
    let res = ""
    for (const item of args) {
      res += typeof item == "string" ? item : JSON.stringify(item)
    }
    console.error(res)
  }

  private _genPrefix() {
    return formatTime() + (this.prefix ? ` [${this.prefix}]` : "")
  }
}

export const logger = new Logger({
  level: LogLevel.DEBUG,
  prefix: "editor",
})

export default Logger
