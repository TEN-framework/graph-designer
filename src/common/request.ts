import { message } from "antd"
import { camelToSnake } from "./utils"
import {
  IConnection,
  IExtension,
  ICompatibleConnection,
  IGraphData,
} from "@/types"
import { DEFAULT_EXTENTION_GROUP, API_URL } from "./constant"

export const apiGetVersion = async () => {
  const res = await fetch(`${API_URL}/version`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get version")
  }
  return data?.data ?? {}
}

export const apiGetInstalledExtension = async () => {
  const res = await fetch(`${API_URL}/addons/extensions`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get installed extensions")
  }
  let arr = data?.data ?? []

  return arr.map((item: IExtension) => {
    return {
      ...item,
      addon: item.name,
      app: "localhost",
      extension_group: DEFAULT_EXTENTION_GROUP,
      property: null,
    }
  })
}

export const apiAllGetGraph = async () => {
  const res = await fetch(`${API_URL}/graphs`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get all graphs")
  }
  return data?.data ?? []
}

export const apiGetGraphExtension = async (
  graphName: string,
): Promise<IExtension[]> => {
  const res = await fetch(`${API_URL}/graphs/${graphName}/extensions`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get graph extensions")
  }
  return (data?.data ?? []) as IExtension[]
}

export const apiGetGraphConnection = async (
  graphName: string,
): Promise<IConnection[]> => {
  const res = await fetch(`${API_URL}/graphs/${graphName}/connections`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get graph connections")
  }
  return (data?.data ?? []) as IConnection[]
}

export const apiQueryCompatibleMessage = async (
  options: ICompatibleConnection,
): Promise<ICompatibleConnection[]> => {
  const res = await fetch(`${API_URL}/messages/compatible`, {
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
    method: "POST",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error(data.message)
  }
  let arr = data?.data ?? []
  return arr as ICompatibleConnection[]
}

export const apiUpdateGraph = async (
  graphName: string,
  graphData: IGraphData,
) => {
  const res = await fetch(`${API_URL}/graphs/${graphName}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(graphData),
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to update graph")
  }
  return data
}

export const apiSaveGraph = async () => {
  const res = await fetch(`${API_URL}/manifest`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to save graph")
  }
  return data
}
