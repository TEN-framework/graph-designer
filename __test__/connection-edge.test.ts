import { expect, test, describe } from 'vitest'
import { connectionsToEdges, MOCK_CONNECTIONS, extensionsToNodes, MOCK_EXTENTIONS, edgesToConnections } from "@/common"
import type { CustomEdge } from "@/types"

let edges: CustomEdge[]
const nodes = extensionsToNodes(MOCK_EXTENTIONS)

describe("connection and edge convert each other", () => {

  test('connections to edges', () => {
    edges = connectionsToEdges(MOCK_CONNECTIONS, nodes)
    expect(edges).toMatchSnapshot();
  })

  test('edges to connections', () => {
    const connections = edgesToConnections(edges, nodes)
    expect(connections).toMatchSnapshot();
  })


})

