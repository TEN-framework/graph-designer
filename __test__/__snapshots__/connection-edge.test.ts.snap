// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connection and edge convert each other > connections to edges 1`] = `
[
  {
    "data": {
      "dataType": "data",
    },
    "id": "1",
    "label": "data",
    "source": "1",
    "sourceHandle": "text_data",
    "target": "2",
    "targetHandle": "text_data",
  },
  {
    "data": {
      "dataType": "data",
    },
    "id": "2",
    "label": "data",
    "source": "1",
    "sourceHandle": "text_data",
    "target": "3",
    "targetHandle": "text_data",
  },
  {
    "data": {
      "dataType": "data",
    },
    "id": "3",
    "label": "data",
    "source": "1",
    "sourceHandle": "text_data",
    "target": "5",
    "targetHandle": "text_data",
  },
  {
    "data": {
      "dataType": "cmd",
    },
    "id": "4",
    "label": "cmd",
    "source": "3",
    "sourceHandle": "flush",
    "target": "4",
    "targetHandle": "flush",
  },
  {
    "data": {
      "dataType": "cmd",
    },
    "id": "5",
    "label": "cmd",
    "source": "4",
    "sourceHandle": "flush",
    "target": "1",
    "targetHandle": "flush",
  },
  {
    "data": {
      "dataType": "data",
    },
    "id": "6",
    "label": "data",
    "source": "5",
    "sourceHandle": "data",
    "target": "1",
    "targetHandle": "data",
  },
  {
    "data": {
      "dataType": "cmd",
    },
    "id": "7",
    "label": "cmd",
    "source": "2",
    "sourceHandle": "flush",
    "target": "3",
    "targetHandle": "flush",
  },
]
`;

exports[`connection and edge convert each other > edges to connections 1`] = `
[
  {
    "app": "localhost",
    "data": [
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "interrupt_detector",
            "extension_group": "default",
          },
        ],
        "name": "text_data",
      },
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "openai_chatgpt",
            "extension_group": "chatgpt",
          },
        ],
        "name": "text_data",
      },
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "chat_transcriber",
            "extension_group": "transcriber",
          },
        ],
        "name": "text_data",
      },
    ],
    "extension": "agora_rtc",
    "extension_group": "default",
  },
  {
    "app": "localhost",
    "cmd": [
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "azure_tts",
            "extension_group": "tts",
          },
        ],
        "name": "flush",
      },
    ],
    "extension": "openai_chatgpt",
    "extension_group": "chatgpt",
  },
  {
    "app": "localhost",
    "cmd": [
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "agora_rtc",
            "extension_group": "default",
          },
        ],
        "name": "flush",
      },
    ],
    "extension": "azure_tts",
    "extension_group": "tts",
  },
  {
    "app": "localhost",
    "data": [
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "agora_rtc",
            "extension_group": "default",
          },
        ],
        "name": "data",
      },
    ],
    "extension": "chat_transcriber",
    "extension_group": "transcriber",
  },
  {
    "app": "localhost",
    "cmd": [
      {
        "dest": [
          {
            "app": "localhost",
            "extension": "openai_chatgpt",
            "extension_group": "chatgpt",
          },
        ],
        "name": "flush",
      },
    ],
    "extension": "interrupt_detector",
    "extension_group": "default",
  },
]
`;
